src/utils/source.js

```diff
@@ -122,32 +122,39 @@ const contentTypeModeMap = {
  */
 
 function getMode(source: Source) {
-  if (!source.text) {
+  const { contentType, text } = source;
+
+  if (!text) {
     return { name: "text" };
   }
 
-  if (!source.contentType) {
-    // Use HTML mode for files in which the first non whitespace
-    // character is `<` regardless of extension.
-    const name = source.text.match(/^\s*</) ? "htmlmixed" : "text";
-    return { name };
+  // Use HTML mode for files in which the first non whitespace
+  // character is `<` regardless of extension.
+  const isHTMLLike = text.match(/^\s*</);
+  if (!contentType) {
+    if (isHTMLLike) {
+      return { name: "htmlmixed" };
+    }
+    return { name: "text" };
   }
 
-  const { contentType, text } = source;
-
   // // @flow or /* @flow */
   if (text.match(/^\s*(\/\/ @flow|\/\* @flow \*\/)/)) {
     return contentTypeModeMap["text/typescript"];
   }
 
-  if (/script|elm|jsx|clojure|wasm/.test(contentType)) {
+  if (/script|elm|jsx|clojure|wasm|html/.test(contentType)) {
     if (contentType in contentTypeModeMap) {
       return contentTypeModeMap[contentType];
     }
 
     return contentTypeModeMap["text/javascript"];
   }
 
+  if (isHTMLLike) {
```

Comment:
the check for `if (!contentType)` has the advantage of an early exit -- we don't perform the other checks, which do rely on content type. So i think we can get rid of lines 154 - 157. If we have a file that is html-like, and it has a contentType, it will be caught by line 146. if all checks fail, we return `{ name: "text" }`

we can also simplify the if statements in the block on lines 134-138

You have a better view of whats going on though, let me know if I miss interpreted!